File: CodeWars/Mathematics/TankTruck/TankTruck.rs
use std::f64;

fn tank_vol(h: i32, d: i32, vt: i32) -> i32 {
    let h = h as f64;
    let r = d as f64/2.0;
    let vt = vt as f64;

    let sector_angle = 2.0 * (1.0-(h/r)).acos();
    let segment_area = (r*r * (sector_angle - sector_angle.sin()))/2.0;
    let cylinder_length = vt / (r*r * f64::consts::PI);

    (segment_area * cylinder_length).floor() as i32
}

fn dotest(h: i32, d: i32, vt: i32, exp: i32) -> () {
    assert_eq!(tank_vol(h, d, vt), exp)
}

fn basics_tank_vol() {
    dotest(5, 7, 3848, 2940);
    dotest(5, 7, 3848, 2940);
    dotest(2, 7, 3848, 907);
    dotest(2, 8, 5026, 982);
    dotest(4, 9, 6361, 2731);
}

fn main() {
    basics_tank_vol();
}

File: CodeWars/Mathematics/TankTruck/TankTruck.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>

int tankVol(int h, int d, int vt) {
    double height = h;
    double radius = d/2.0;
    double velocity = vt;

    double sector_angle = 2.0 * acos((1.0-(height/radius)));
    double segment_area = (radius*radius * (sector_angle - sin(sector_angle)))/2.0;
    double cylinder_length = velocity / (radius*radius * M_PI);

    return floor(segment_area * cylinder_length);
}

void dotest(int h, int d, int vt, int expr) {
    int act = tankVol(h, d, vt);
    if(act != expr)
        printf("Error. Expected %d but got %d\n", expr, act);
    assert(act == expr);
}

int main() {
    dotest(5, 7, 3848, 2940);
    dotest(2, 7, 3848, 907);
    dotest(3, 6, 3500, 1750);
    dotest(3, 6, 3501, 1750);
}

